hasktorch/src/Torch/Random.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Torch.Random
+  ( mkGenerator
+  , Generator
+  , randn
+  , randn'
+  , rand
+  , rand'
+  , randint
+  , randint'
+  , normal
+  , normal'
+  ) where
+
+import ATen.Cast
+import ATen.Class (Castable(..))
+import qualified ATen.Const as ATen
+import qualified ATen.Type as ATen
+import Data.Word
+import qualified ATen.Managed.Type.Generator as ATen
+import qualified Torch.Managed.Native as LibTorch
+import Torch.Device
+import Torch.TensorOptions
+import Torch.Tensor
+import Foreign.ForeignPtr
+import System.IO.Unsafe
+import           Control.Concurrent
+import           Control.Concurrent.STM
+import           Control.Monad.IO.Class
+import           Control.Monad.STM
+
+instance Show (TVar (Maybe (ForeignPtr ATen.Generator))) where
+  show _ = "_"
+
+data Generator
+  = Generator
+  { device :: Device
+  , seed :: Word64
+  , generator :: TVar (Maybe (ForeignPtr ATen.Generator))
+  }
+  deriving (Eq,Show)
+
+mkGenerator :: Device -> Word64 -> IO Generator
+mkGenerator device seed =
+  case device of
+    Device CPU _ -> do
+      genPtr <- ATen.newCPUGenerator seed
+      genenerator <- newTVarIO (Just genPtr)
+      return $ Generator device seed genenerator
+    Device CUDA idx -> do
+      genPtr <- ATen.newCUDAGenerator (fromIntegral idx)
+      ATen.generator_set_current_seed genPtr seed
+      genenerator <- newTVarIO (Just genPtr)
+      return $ Generator device seed genenerator
+
+type RandomGenFunc = ForeignPtr ATen.IntArray -> ForeignPtr ATen.Generator -> ForeignPtr ATen.TensorOptions -> IO (ForeignPtr ATen.Tensor)
+
+generatorFactory :: RandomGenFunc -> [Int] -> TensorOptions -> Generator -> (Tensor,Generator)
+generatorFactory func size options Generator{..} =
+  unsafePerformIO $ do
+    mGenerator <- atomically $ do
+      v <- readTVar generator
+      case v of
+        Just v' -> do
+          writeTVar generator Nothing
+          return $ Just v'
+        Nothing -> return Nothing
+    genPtr <- case mGenerator of
+                Just gen -> return gen
+                Nothing -> case device of
+                  Device CPU _ -> ATen.newCPUGenerator seed
+                  Device CUDA idx -> do
+                    gen <- ATen.newCUDAGenerator (fromIntegral idx)
+                    ATen.generator_set_current_seed gen seed
+                    return gen
+    tensor <- (cast3 func) size genPtr options
+    nextSeed <- ATen.generator_current_seed genPtr
+    nextGenenerator <- newTVarIO (Just genPtr)
+    return (tensor,Generator device nextSeed nextGenenerator)
```

Comment:
I don't think we should do this. PyTorch RNG infrastructure is not designed around efficient generator copying and it's very likely that the performance will be terrible. IIRC We might be using some PRNGs with huge states.

Reply:
@junjihashimoto hm, can you confirm that the ATen generator is actually copied? To me it looks like you are just mutating a TVar with maybe an ATen generator in it, allocating a new one only at first use

Reply:
@apaszke @tscholak 

This data with TVar manages ownership of ATen generator.
When the data has ownership, copy does not occur.
When the data does not have ownership, new ATen generator is created by seed.

Let me show you two examples.

Case1:
```
rng0 <- mkGenerator ..
(t1,rng1) = rand' [] rng0
(t2,rng2) = rand' [] rng1
```
In this case, copy does not occur.
rng0's ownership moves to rng1, next rng1 moves to rng2.

Case2:
rng0 <- mkGenerator ..
(t1,rng1) = rand' [] rng0
(t2,rng2) = rand' [] rng0

In this case, copy occurs.
rng0's ownership moves to rng1, next rng0 does not have ownership, and rng2 is created from seed.

Reply:
Oh I see. However, I'm not convinced that `current_seed` gives you a seed that would yield an equivalent generator. IIRC it just returns the seed that it was initialized with, so it will produce a constant stream as a result? Would need to check the PyTorch code though.

Reply:
> so it will produce a constant stream as a result?

Yes.
This randn's test reproduces the same stream with previous called stream.
https://github.com/hasktorch/hasktorch/pull/250/files#diff-1ad63067a684ef689650d73894fc7038R19

> Would need to check the PyTorch code though.

I will do it later.